### DISCUSSION TITLE
Bitrate Allocation of Latent Representations

### DISCUSSION THREAD

**Discussion by rezafuru** [yoshitomo-matsubara/sc2-benchmark#104] [Q&A] [answered]
Hi,

I've really liked your work on compression for Split Computing. However, it is quite difficult to wrap my head around all the neural compression concepts. 

One thing in particular, I really struggle to understand is how you created the visualization of **Bitrate Allocation of Latent Representations** (Figure 11 in SC2 and Figure 7 in Supervised Compression for Resource-Constrained Edge Computing Systems)

Specifically, I'm not sure, how you can measure the allocated bitrate by using the latent representation. The way I understood this is that, during training, you gain compression rate by maximizing the mutual information between the latent variable **z** and the intermediate features **h**.  So how do you derive the spatial information (i.e. which dimension in z corresponds to which pixels in the original information) out of the latent representation?

Ideally, if it is not too much work, you could upload the code of the visualization in this repository. 

Best
Alireza

**Comment by yoshitomo-matsubara** [comment, marked as answer]
Hi @rezafuru ,

Thank you for having an interest in our papers!

We computed bitrates (using output from entropy bottleneck) regarding the latent variable $\mathbf{z}$ and visualize the pixel-wise bitrate info as "Bitrate Allocation of Latent Representations".
Here is an minimal example code for reference.

```python
compression_model = factorized_prior # (in compressai package)
compression_model.eval()
compression_model.update()

x = input_image_tensor # assume batch size of 1 i.e. [1, 3, height, width]
latent = compression_model.g_a(x)
latent_hat, likelihoods = compression_model.entropy_bottleneck(latent)
likelihood = likelihoods[0].detach().cpu().numpy()
pixelwise_bitrates = -np.log2(likelihood).sum(0)
pixelwise_bitrates_norm = (pixelwise_bitrates - pixelwise_bitrates.mean()) / np.std(pixelwise_bitrates)
```

If this answers the question, choose "mark as answer" and close this discussion.
Feel free to open a new discussion when you have different questions.

**Reply by rezafuru** [reply]
Thank you very much for the answer, and again thank you very much for putting so much effort into open sourcing all the code of your papers! It is helping immensely

Best
Alireza
